File: Client/Core/src/main/java/me/footlights/core/crypto/Keychain.java
package me.footlights.core.crypto;

import java.io.*;
import java.security.*;
import java.security.cert.*;
import java.util.Arrays;
import java.util.Enumeration;
import java.util.Map;

import com.google.common.base.Function;
import com.google.common.collect.Maps;

import me.footlights.core.Config;


/** Stores crypto keys. */
public class Keychain
{
	public Map<String,PublicKey> publicKeys()
	{
		return Maps.transformValues(privateKeys, PRIVATE_TO_PUBLIC);
	}

	public void store(PrivateKey key)
	{
		String name = key.getFingerprint().encode();

		if (privateKeys.containsKey(name))
			assert Arrays.equals(
					key.key.getPrivateKey().getEncoded(),
					privateKeys.get(name).key.getPrivateKey().getEncoded());

		privateKeys.put(name, key);
	}

	public void store(SecretKey key)
	{
		String name = key.getFingerprint().encode();

		if (secretKeys.containsKey(name))
			assert Arrays.equals(
					key.keySpec.getEncoded(),
					secretKeys.get(name).keySpec.getEncoded());

		secretKeys.put(name, key);
	}



	/** Merge a KeyStore file into this Keychain. */
	public void importKeystoreFile(InputStream input)
		throws CertificateException, IOException, KeyStoreException,
		       NoSuchAlgorithmException, UnrecoverableEntryException
	{
		importKeystoreFile(input, Config.getInstance().get("crypto.keystore.type"));
	}

	/** Merge a KeyStore file into this Keychain. */
	public void importKeystoreFile(InputStream input, String type)
		throws CertificateException, IOException, KeyStoreException, NoSuchAlgorithmException,
		       UnrecoverableEntryException
	{
		final String password = getPassword();

		KeyStore store = KeyStore.getInstance(type);
		store.load(input, password.toCharArray());

		final KeyStore.ProtectionParameter protection =
			new KeyStore.PasswordProtection(password.toCharArray());

		Enumeration<String> aliases = store.aliases();
		while (aliases.hasMoreElements())
		{
			final String alias = aliases.nextElement();
			final String[] parts = alias.split(ALIAS_SEPARATOR);
			if (parts.length != 3)
				throw new KeyStoreException("Expected 'type:algorithm:hash', got '" + alias + "'");

			final String entryType = parts[0].toUpperCase();
			final String fingerprintAlgorithm = parts[1];

			EntryType t;
			try { t = EntryType.valueOf(entryType); }
			catch (IllegalArgumentException e)
			{
				throw new KeyStoreException("Invalid keystore entry type '" + parts[0] + "'");
			}

			KeyStore.Entry entry = store.getEntry(alias, protection);
			switch (t)
			{
				case PRIVATE:
					KeyStore.PrivateKeyEntry privateKey = (KeyStore.PrivateKeyEntry) entry;
					store(
						new PrivateKey(
							privateKey,
							Fingerprint.newBuilder()
								.setAlgorithm(fingerprintAlgorithm)
								.setContent(privateKey.getCertificate().getPublicKey().getEncoded())
								.build()));
					break;

				case SECRET:
					javax.crypto.SecretKey secretKey = ((KeyStore.SecretKeyEntry) entry).getSecretKey();
					store(
						SecretKey.newGenerator()
							.setCryptoAlgorithm(secretKey.getAlgorithm())
							.setFingerprintAlgorithm(fingerprintAlgorithm)
							.setBytes(secretKey.getEncoded())
							.generate());
					break;
			}
		}
	}


	/** Save a Keychain to a KeyStore file. */
	void exportKeystoreFile(OutputStream out)
		throws CertificateException, IOException, KeyStoreException, NoSuchAlgorithmException,
		       UnrecoverableEntryException
   {
		exportKeystoreFile(out, Config.getInstance().get("crypto.keystore.type"));
   }

	/** Save a Keychain to a KeyStore file. */
	void exportKeystoreFile(OutputStream out, String type)
		throws CertificateException, IOException, KeyStoreException, NoSuchAlgorithmException,
		       UnrecoverableEntryException
	{
		final String password = getPassword();

		KeyStore store = KeyStore.getInstance(type);
		store.load(null, password.toCharArray());

		final KeyStore.ProtectionParameter protection =
			new KeyStore.PasswordProtection(password.toCharArray());

		for (PrivateKey key : privateKeys.values())
			store.setEntry("private:" + key.getFingerprint().encoded(), key.key, protection);

		for (final SecretKey secret : secretKeys.values())
			store.setEntry(
					"secret:" + secret.getFingerprint().encoded(),
					new KeyStore.SecretKeyEntry(secret.keySpec),
					protection);

		store.store(out, password.toCharArray());
	}




	/** Kinds of things that we store in a Java Keystore. */
	private enum EntryType
	{
		/** A certificate. */
		CERT,

		/** A private key for a public/private keypair. */
		PRIVATE,

		/** A secret, symmetric key. */
		SECRET,
	}


	/** Retrieve the keystore password from somewhere trustworthy (the user?). */
	private final String getPassword() { return "fubar"; }


	private static final String	ALIAS_SEPARATOR  = ":";
	
	private static final Function<PrivateKey, PublicKey> PRIVATE_TO_PUBLIC =
		new Function<PrivateKey, PublicKey>()
		{
			public PublicKey apply(PrivateKey priv) { return priv.publicKey(); }
		};


	/** Public/private keypairs. */
	private final Map<String, PrivateKey> privateKeys = Maps.newHashMap();

	/** Secret keys for decrypting blocks */
	private final Map<String, SecretKey> secretKeys = Maps.newHashMap();
}

File: Client/Core/src/main/scala/me/footlights/core/crypto/keychain.scala
/*
 * Copyright 2012 Jonathan Anderson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.io.{ByteArrayOutputStream, IOException}
import java.net.URI
import java.nio.ByteBuffer
import java.nio.channels.{Channels, ReadableByteChannel, WritableByteChannel}

import java.security.KeyStore
import java.security.cert.{Certificate,CertificateException}

import java.util.logging.Logger

import scala.collection.JavaConversions._

import me.footlights.core
import me.footlights.core.IO._
import me.footlights.core.data
import me.footlights.core.data.FormatException


package me.footlights.core.crypto {


/**
 * Stores cryptographic material.
 *
 * A Keychain can store two kinds of cryptographic material:
 * 1. symmetric keys, represented in {@link Link} objects which have authority to decrypt things and
 * 2. asymmetric keys, exposed as {@link SingingIdentity} objects have authority to can sign things.
 */
trait Keychain extends core.HasBytes {
	private[crypto] def links:Map[Fingerprint,Link]
	private[crypto] def serialized:ByteBuffer

	/** Concatenate with a link to an {@link EncryptedBlock}. */
	def + (link:Link): Keychain = new ImmutableKeychain(links + (link.fingerprint -> link))

	def ++ (k:Keychain): Keychain = new ImmutableKeychain(links ++ k.links)


	/** Get the {@link Link} which is capable of decrypting the named ciphertext. */
	def getLink(name:Fingerprint) = links get name

	/** Get a serialized representation which can be saved in the CAS. */
	override def getBytes = serialized.asReadOnlyBuffer 

	// Object overrides
	override def toString = "Keychain { %s }" format links
	override def equals(other:Any) = other match {
		case o:Keychain => links equals o.links
		case _ => false
	}


	/** Export to a Java {@link KeyStore} (e.g. the root {@link Keychain} on local disk). */
	private[core] def exportKeyStore(channel:WritableByteChannel,
			keystoreType:String = Keychain.defaultKeystoreType(),
			password:String = Keychain.getPassword()) = {
		log fine "Exporting to Java KeyStore..."

		val store = KeyStore.getInstance(keystoreType)
		store.load(null, password.toCharArray)    // completes KeyStore initialization (bad API!)

		// Protect keys with the same password as the keystore itself (common practice).
		val protection = new KeyStore.PasswordProtection(password.toCharArray)

		for ((fingerprint, link) <- links)
			store.setEntry(fingerprint.encode, new KeyStore.SecretKeyEntry(link.key.keySpec),
					protection)

		log fine { "Saved %d symmetric keys to Java KeyStore" format links.size }

		store.store(Channels newOutputStream channel, password toCharArray)
		store
	}

	private val log = Logger getLogger classOf[Keychain].getCanonicalName
}

/**
 * A mutable data structure which holds a variable, immutable {@link Keychain} and (optionally)
 * notifies an observer when that keychain changes.
 */
class MutableKeychain (private var keychain:Keychain, notify:Keychain => Unit = (_ => Unit))
		extends Keychain {

	def store(link:Link) = synchronized {
		keychain += link
		notify(keychain)
		this
	}

	private[crypto] def links() = keychain.links
	private[crypto] def serialized() = keychain.serialized
}


/** An immutable {@link Keychain} which can be concatenated with other things. */
class ImmutableKeychain private[crypto](
		private[crypto] val links:Map[Fingerprint,Link]) extends Keychain {

	private[crypto] lazy override val serialized =  {
		val out = new ByteArrayOutputStream()
		Keychain.serialize(this, Channels.newChannel(out))
		ByteBuffer wrap out.toByteArray
	}

	private val log = Logger getLogger classOf[Keychain].getCanonicalName
}


object Keychain {
	def apply(links:Iterable[(Fingerprint, Link)]): Keychain = apply(Map() ++ links)
	def apply(links:Map[Fingerprint,Link] = Map()): Keychain = new ImmutableKeychain(links)


	// TODO: return Either[Exception, Keychain]
	def parse(bytes:ByteBuffer): Either[Exception, Keychain] = {
		val magic = new Array[Byte](Magic.length)
		bytes get magic
		if (magic.deep != Magic.deep)
			Left(new FormatException("Invalid keychain magic: " + magic.array.toList))

		else try {
			val count = bytes getInt

			val links = for (i <- 0 until count) yield {
				val (namelen, keylen) = (bytes getInt, bytes getInt)
				val fingerprint = new Array[Byte](namelen)
				bytes get fingerprint

				val key = new Array[Byte](keylen)
				bytes get key

				val name = Fingerprint decode { new String(fingerprint) }
				val secret = SecretKey parse { new URI(new String(key)) }

				name -> (secret.createLinkBuilder setFingerprint name setKey secret build)
			}

			Right(apply(links))
		} catch {
			case ex:Exception => Left(ex)
		}
	}


	private[crypto] def serialize(keychain:Keychain, out:WritableByteChannel) = {
		out << Magic << keychain.links.size

		for ((fingerprint,link) <- keychain.links) {
			val name = fingerprint.encode
			val secret = link.key.toUri.toString

			out << name.length << secret.length
			out << name.getBytes << secret.getBytes
		}
	}


	def importKeyStore(channel:ReadableByteChannel,
			storeType:String = defaultKeystoreType(),
			password:String = getPassword()) = {

		val rawPassword = password.toCharArray

		val store = KeyStore getInstance storeType
		store.load(if (channel.isOpen) Channels newInputStream channel else null, rawPassword)
		log fine "Loaded %d KeyStore entries".format(store.size)

		val links = for (alias <- store.aliases) yield alias match {
			case Fingerprint(fingerprint) =>
				val secret = store.getKey(alias, rawPassword).asInstanceOf[javax.crypto.SecretKey]

				val key = SecretKey.newGenerator
					.setAlgorithm(secret.getAlgorithm)
					.setFingerprintAlgorithm(fingerprint.getAlgorithm.getAlgorithm)
					.setBytes(secret.getEncoded)
					.generate
				val link = key.createLinkBuilder
					.setFingerprint(fingerprint)
					.build

				Some(fingerprint -> link)

			case _ =>
				log warning "Invalid KeyStore entry '%s'".format(alias)
				None
		}

		apply(links.flatten toIterable)
	}


	/** Retrieve the keystore password from somewhere trustworthy (the user?). */
	private def getPassword() = "fubar"

	private lazy val prefs = core.Preferences.getDefaultPreferences
	private def defaultKeystoreType() = prefs getString "crypto.keystore.type" getOrElse {
		throw new core.ConfigurationError("No default keystore type configured")
	}

	/** Magic hexword: "Foot keys". */
	private val Magic = List(0xF0, 0x07, 0x6E, 0x75) map { _.toByte } toArray

	private val log = Logger getLogger Keychain.getClass.getCanonicalName
}

}
